Source/Search/Mixins/AdvancedSearchProviderMixin.lua

```diff
@@ -0,0 +1,143 @@
+    strsplit( delimiter, searchString )
+  --TODO Convert filterKey to filters/itemClassFilters in the browse query
+
+  --TODO Check for searching the whole AH by mistake (which causes a DC)
+
+  minLevel = tonumber( minLevel )
+  maxLevel = tonumber( maxLevel )
+  minItemLevel = tonumber( minItemLevel )
+  maxItemLevel = tonumber( maxItemLevel )
+
+  if minLevel == 0 then
+    minLevel = nil
+  end
+
+  if maxLevel == 0 then
+    maxLevel = nil
+  end
+
+  if minItemLevel == 0 then
+    minItemLevel = nil
+  end
+
+  if maxItemLevel == 0 then
+    maxItemLevel = nil
+  end
+
+  return {
+    query = {
+      searchString = queryString,
+      minLevel = minLevel,
+      maxLevel = maxLevel,
+      filters = {},
+      itemClassFilters = {},
+      sorts = {}
+    },
+    extraFilters = {
+      minItemLevel = minItemLevel,
+      maxItemLevel = maxItemLevel
+    }
+  }
+end
+
+function AuctionatorAdvancedSearchProviderMixin:CreateSearchTerm(term)
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:CreateSearchTerm()", term)
+  if IsCompoundSearch(term) then
+    return ParseCompoundSearch(term)
+  else
+    return  {
+      query = {
+        searchString = term,
+        filters = {},
+        itemClassFilters = {},
+        sorts = {},
+      },
+      extraFilters = {}
+    }
+  end
+end
+
+function AuctionatorAdvancedSearchProviderMixin:GetSearchProvider()
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:GetSearchProvider()")
+
+  return function(searchTerm)
+    print(searchTerm.query.searchString)
+    C_AuctionHouse.SendBrowseQuery(searchTerm.query)
+    self.filter = searchTerm.extraFilters
+  end
+end
+
+function AuctionatorAdvancedSearchProviderMixin:OnSearchEventReceived(eventName, ...)
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:OnSearchEventReceived()", eventName, ...)
+
+  if eventName == "AUCTION_HOUSE_BROWSE_RESULTS_UPDATED" then
+    self:ProcessSearchResults(C_AuctionHouse.GetBrowseResults())
+  elseif eventName == "AUCTION_HOUSE_BROWSE_RESULTS_ADDED" then
+    self:ProcessSearchResults(...)
+  elseif eventName == "AUCTION_HOUSE_BROWSE_FAILURE" then
+    AuctionHouseFrame.BrowseResultsFrame.ItemList:SetCustomError(
+      RED_FONT_COLOR:WrapTextInColorCode(ERR_AUCTION_DATABASE_ERROR)
+    )
+  end
+end
+
+function AuctionatorAdvancedSearchProviderMixin:ProcessSearchResults(addedResults)
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:ProcessSearchResults()")
+
+  local results = {}
+
+  for index = 1, #addedResults do
+    if self:FilterItemKey(addedResults[index].itemKey) then
+      table.insert(results, addedResults[index].itemKey)
+    else
+      print("filter failed")
+    end
+  end
+
+  self:AddResults(results)
+end
+
+function AuctionatorAdvancedSearchProviderMixin:FilterItemKey(itemKey)
+  if itemKey.itemLevel ~= nil and self.filter ~= nil then
+    return (not self.filter.minItemLevel or self.filter.minItemLevel<=itemKey.itemLevel)
```

Comment:
This boolean statement is confusing to me... Is it:
```lua
self.filter.minItemLevel ~= nil and self.filter.minItemLevel <= itemKey.itemLevel and
self.filter.maxItemLevel ~= nil and self.filter.maxItemLevel >= itemKey.itemLevel
```

Reply:
Oh, whoops, didn't fix that yet...